core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -0,0 +1,533 @@
+use std::{ops, time::Instant};
+
+use anyhow::Context;
+use zksync_dal::CoreDal;
+use zksync_multivm::{
+    interface::{TransactionExecutionMetrics, VmExecutionResultAndLogs},
+    utils::{
+        adjust_pubdata_price_for_tx, derive_base_fee_and_gas_per_pubdata, derive_overhead,
+        get_max_batch_gas_limit,
+    },
+    zk_evm_latest::ethereum_types::H256,
+};
+use zksync_system_constants::MAX_L2_TX_GAS_LIMIT;
+use zksync_types::{
+    api::state_override::StateOverride, fee::Fee, fee_model::BatchFeeInput, get_code_key,
+    ExecuteTransactionCommon, PackedEthSignature, ProtocolVersionId, Transaction,
+};
+
+use super::{result::ApiCallResult, SubmitTxError, TxSender};
+use crate::execution_sandbox::{BlockArgs, SandboxAction, VmPermit, SANDBOX_METRICS};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum BinarySearchKind {
+    /// Full binary search.
+    Full,
+    /// Binary search with an optimized initial pivot.
+    #[allow(dead_code)] // FIXME
+    Optimized,
+}
+
+impl TxSender {
+    #[tracing::instrument(level = "debug", skip_all, fields(
+        initiator = ?tx.initiator_account(),
+        nonce = ?tx.nonce(),
+    ))]
+    pub async fn get_txs_fee_in_wei(
+        &self,
+        tx: Transaction,
+        estimated_fee_scale_factor: f64,
+        acceptable_overestimation: u64,
+        state_override: Option<StateOverride>,
+        kind: BinarySearchKind,
+    ) -> Result<Fee, SubmitTxError> {
+        let estimation_started_at = Instant::now();
+        let mut estimator = GasEstimator::new(self, tx, state_override).await?;
+        estimator.adjust_transaction_fee();
+
+        let initial_estimate = estimator.initialize().await?;
+        tracing::trace!(
+            "preparation took {:?}, starting binary search",
+            estimation_started_at.elapsed()
+        );
+
+        let optimized_lower_bound = initial_estimate.lower_gas_bound_without_overhead();
+        // Perform an initial search iteration with the pivot slightly greater than `gas_used` to account for 63/64 rule for far calls etc.
+        // If the transaction succeeds, it will discard most of the search space at once.
+        let optimistic_gas_limit = initial_estimate.optimistic_gas_limit_without_overhead();
+
+        let (bounds, initial_pivot) = match kind {
+            BinarySearchKind::Full => {
+                let lower_bound = initial_estimate.gas_charged_for_pubdata;
+                let upper_bound = MAX_L2_TX_GAS_LIMIT;
```

Comment:
`upper_bound` should be `initial_estimate.gas_charged_for_pubdata + MAX_L2_TX_GAS_LIMIT`, that's how it works in old code, no? `MAX_L2_TX_GAS_LIMIT`'s name is somewhat wrong, in fact it means `MAX_L2_TX_COMPUTATIONAL_GAS_LIMIT`

Reply:
@slowli is it worth it to rename the var? Feel free to keep as is too